### PR TITLE
Add `MultiNestedTensor`

### PR DESCRIPTION
**Pull request by weihua916**
* Add basic functionality of `MultiNestedTensor` that supports Pytorch Tensor of shape `(num_rows, num_cols, *)`, where the last dimension can have varying sizes.

Basic usage is that
```python
# Prepare tensor_mat: List[List[Tensor]]
multi_nested_tensor = MultiNestedTensor.from_tensor_mat(tensor_mat)
# Get (i,j)-th element in Tensor
multi_nested_tensor[i, j]
# Get i-th row in `MultiNestedTensor`
multi_nested_tensor[i]
# Row index select
multi_nested_tensor[[3, 4, 5]]
```

Column index select, row slicing/arange will be supported in later PR.

### CODE REVIEW

File: torch_frame/data/multi_nested_tensor.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
+import copy
+from typing import Any, Callable, Dict, List, Tuple, Union
+
+import torch
+from torch import Tensor
+
+
+class MultiNestedTensor:
+    r"""A PyTorch Tensor-based data structure that stores
+    :obj:`[num_rows, num_cols, *]` , where the size of last dimension can be
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:10]:
```suggestion
    r"""A PyTorch tensor-based data structure that stores
    :obj:`[num_rows, num_cols, *]`, where the size of last dimension can be
```

File: torch_frame/data/multi_nested_tensor.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    different for different data/column. Internally, we store the object in an
+    efficient flattened format: :obj:`(value, offset)`, where the element
+    at :obj:`(i, j)` is accessed by
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py]: by what?

File: torch_frame/data/multi_nested_tensor.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    def from_tensor_mat(
+        cls,
+        tensor_mat: List[List[Tensor]],
+    ) -> 'MultiNestedTensor':
+        r"""Construct :class:`MultiNestedTensor` object from
+        :obj:`tensor_mat`.
+
+        Args:
+            tensor_mat List[List[Tensor]]: A dictionary of
+                matrix of PyTorch Tensors. :obj:`tensor_mat[i][j]` contains
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:47]: Type hint doesn't match with dictionary.

File: torch_frame/data/multi_nested_tensor.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        name += f"(num_rows: {self.num_rows}, num_cols: {self.num_cols})"
+        return name
+
+    def __copy__(self) -> 'MultiNestedTensor':
+        out = self.__class__.__new__(self.__class__)
+        for key, value in self.__dict__.items():
+            out.__dict__[key] = value
+
+        out.values = copy.copy(out.values)
+        out.offset = copy.copy(out.offset)
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:87]: What does this copy do?

File: torch_frame/data/multi_nested_tensor.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            # Get an element of (i, j). Returns Tensor
+            assert len(index) == 2
+            if index[0] < 0 or index[0] >= self.num_rows:
+                raise IndexError(
+                    f"Index out-of-bounds. The first element of {index} needs "
+                    f"to be [0, {self.num_rows}]")
+            if index[1] < 0 or index[1] >= self.num_cols:
+                raise IndexError(
+                    f"Index out-of-bounds. The second element of {index} "
+                    f"needs to be [0, {self.num_cols}]")
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:105]: Why would that raise an error? We should support negative indices.

File: torch_frame/data/multi_nested_tensor.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            # Get i-th row. Returns MultiNestedTensor
+            start_idx = index * self.num_cols
+            end_idx = (index + 1) * self.num_cols + 1
+            offset = self.offset[start_idx:end_idx]
+            values = self.values[offset[0]:offset[-1]]
+            offset = offset - offset[0]
+            return MultiNestedTensor(num_rows=1, num_cols=self.num_cols,
+                                     values=values, offset=offset)
+        elif isinstance(index, Tensor) and index.ndim == 1:
+            return self._row_index_selet_helper(index)
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:125]:
```suggestion
            return self._row_index_select_helper(index)
```

File: torch_frame/data/multi_nested_tensor.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                    f"[0, {self.num_rows}]")
+            # Get i-th row. Returns MultiNestedTensor
+            start_idx = index * self.num_cols
+            end_idx = (index + 1) * self.num_cols + 1
+            offset = self.offset[start_idx:end_idx]
+            values = self.values[offset[0]:offset[-1]]
+            offset = offset - offset[0]
+            return MultiNestedTensor(num_rows=1, num_cols=self.num_cols,
+                                     values=values, offset=offset)
+        elif isinstance(index, Tensor) and index.ndim == 1:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:124]: we should just expose `index_select`, just as PyTorch tensors do.

File: torch_frame/data/multi_nested_tensor.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+            return self._row_index_selet_helper(index)
+        elif isinstance(index, List):
+            return self._row_index_selet_helper(torch.LongTensor(index))
+        else:
+            raise RuntimeError("Advanced indexing not supported yet.")
+
+    def _row_index_selet_helper(self, index: Tensor) -> Tensor:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:132]:
```suggestion
    def _row_index_selet_helper(self, index: Tensor, dim: int = 0) -> Tensor:
```

**Review by rusty1s — COMMENTED** [review]
Overall, this is pretty nice. Just a few comments.

File: torch_frame/data/multi_nested_tensor.py (hunk 5 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:105]: Good point.

File: torch_frame/data/multi_nested_tensor.py (hunk 7 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:124]: Sounds good.

File: torch_frame/data/multi_nested_tensor.py [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+            for j in range(num_cols):
+                tensor = tensor_mat[i][j]
+                if not isinstance(tensor, Tensor):
+                    raise RuntimeError(
+                        "The element of tensor_mat must be PyTorch Tensor")
+                values_list.append(tensor)
+                accum_idx += len(tensor)
+                offset_list.append(accum_idx)
+
+        values = torch.cat(values_list)
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:70]: As we discussed, we can do dictionary of `MultiNestedTensor`.

File: torch_frame/data/multi_nested_tensor.py [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+        name += f"(num_rows={self.num_rows}, num_cols={self.num_cols}, "
+        name += f"device={self.values.device})"
+        return name
+
+    def __copy__(self) -> 'MultiNestedTensor':
+        out = self.__class__.__new__(self.__class__)
+        for key, value in self.__dict__.items():
+            out.__dict__[key] = value
+
+        out.values = copy.copy(out.values)
```
[the remark this answers is not in the document]
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:86]: It turns out that `copy.copy` is not necessary.

File: torch_frame/data/multi_nested_tensor.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        name = self.__class__.__name__
+        name += f"(num_rows={self.num_rows}, num_cols={self.num_cols}, "
+        name += f"device={self.values.device})"
+        return name
+
+    def __copy__(self) -> 'MultiNestedTensor':
+        out = self.__class__.__new__(self.__class__)
+        for key, value in self.__dict__.items():
+            out.__dict__[key] = value
+        return out
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:88]: This is the default, so you can drop.

File: torch_frame/data/multi_nested_tensor.py [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        values = torch.cat(values_list)
+        offset = torch.LongTensor(offset_list)
+
+        return cls(num_rows, num_cols, values, offset)
+
+    def __repr__(self) -> str:
+        name = self.__class__.__name__
+        name += f"(num_rows={self.num_rows}, num_cols={self.num_cols}, "
+        name += f"device={self.values.device})"
+        return name
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:82]: A bit awkward, I would just write this as a single multi-line string.

File: torch_frame/data/multi_nested_tensor.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+                                     values=values, offset=offset)
+        elif isinstance(index, Tensor) and index.ndim == 1:
+            return self.index_selet(index, dim=0)
+        elif isinstance(index, List):
+            return self.index_selet(
+                torch.LongTensor(index, device=self.values.device), dim=0)
+        else:
+            raise RuntimeError("Advanced indexing not supported yet.")
+
+    def index_selet(self, index: Tensor, dim: int) -> Tensor:
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:137]:
```suggestion
    def index_select(self, index: Tensor, dim: int) -> Tensor:
```

File: torch_frame/data/multi_nested_tensor.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+            offset = self.offset[start_idx:end_idx]
+            values = self.values[offset[0]:offset[-1]]
+            offset = offset - offset[0]
+            return MultiNestedTensor(num_rows=1, num_cols=self.num_cols,
+                                     values=values, offset=offset)
+        elif isinstance(index, Tensor) and index.ndim == 1:
+            return self.index_selet(index, dim=0)
+        elif isinstance(index, List):
+            return self.index_selet(
+                torch.LongTensor(index, device=self.values.device), dim=0)
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:133]:
```suggestion
                torch.tensor(index, device=self.values.device), dim=0)
```

File: torch_frame/data/multi_nested_tensor.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+        else:
+            raise RuntimeError("Advanced indexing not supported yet.")
+
+    def index_selet(self, index: Tensor, dim: int) -> Tensor:
+        if dim == 0:
+            return self._row_index_select(index)
```
Comment (rusty1s) [inline torch_frame/data/multi_nested_tensor.py:139]:
```suggestion
        if dim == 0 or dim == -3:
            return self._row_index_select(index)
```

**Review by rusty1s — APPROVED** [review]

File: torch_frame/data/multi_nested_tensor.py (hunk 15 above)
Comment (weihua916) [reply torch_frame/data/multi_nested_tensor.py:135]: :)

**Pull request merged**